apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -287,6 +287,11 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
     return packageDescription;
   }
 
+  private _getValidDependencyVersion(dependencyVersion: string): string {
+    const validDependencyVersion: string = dependencyVersion.split('/').pop()!;
+    return semver.valid(validDependencyVersion) ? validDependencyVersion : dependencyVersion.split('_')[0]!;
+  }
+
```

Comment:
The original code was being a bit lazy with `dependencyVersion.split('/').pop()!`, but now that we're expanding the logic, it might be time to beef up the rigor a bit:

- The input format is no longer trivially simple
- This input comes from an unvalidated external file, and that file has already betrayed us once

Anyway the next step up is a `RegExp`, something like this:
```ts
class TheClass {
  // Example: /react-loadable/5.5.0
  private static _pnpm2VersionRegExp: RegExp = /\/([^/_@]+)$/;

  // Example: 5.5.0_react@16.8.6
  private static _pnpm3VersionRegExp = /^([^/_@]+)_/;

   public getValidDependencyVersion(dependencyVersion: string): string {
     let match: RegExpExecArray | null = MyClass._pnpm2VersionRegExp.exec(dependencyVersion);
     if (match) {
       return match[1];
     }
     match = MyClass._pnpm3VersionRegExp.exec(dependencyVersion);
     if (match) {
       return match[1];
     }
     throw new Error(`Invalid version format: "${dependencyVersion}"`)
   }
}
```
(BTW I like to say *"If a function relies on `split()`, it probably has at least one bug"* -- but it's unfair here, since that code predates your PR heheh.)

Reply:
The `semver.valid()` thing seems a bit lax -- it's basically saying _"If it doesn't look like a version, meh, let's pretend we didn't see it."_  Ideally we'd want to have a clear idea about what exactly we're ignoring, but I don't want to hold up this PR, since it's an important fix.  :-)

Reply:
@octogonz I think we should merge this in since it is a blocking issue and open a separate PR for optimizing. I don't particularly like my fix either since even `dependencyVersion.split('/').pop()!` seems to be redundant. With pnpm 2.x it looks like dependency version never comes in as `/react-loadable/5.5.0` I think the slashes get stripped out elsewhere? 
I don't want to depart too much from the original implementation though before I thoroughly test it with 2.x and 3.x and verify that the cases we need to handle are actually just `/react-loadable/5.5.0` and `/react-loadable/5.5.0_react@16.8.6` ( / and _)

Reply:
Sounds good 👍